File: core-app/src/demo-agent.ts
import demoFixtures from '../fixtures/demo-responses.json' with { type: 'json' }
import type { ChatMessage } from './types.js'
import { lastUserMessage } from './session.js'

export type DemoFixtures = {
  helpText: string
  faqAnswers: string[]
  echoTemplate: string
}

export const fixtures: DemoFixtures = demoFixtures

let faqRotation = 0

export function resetFaqRotation(): void {
  faqRotation = 0
}

export function pickFaqAnswer(answers: string[]): string {
  if (answers.length === 0) return ''
  const answer = answers[faqRotation % answers.length]!
  faqRotation += 1
  return answer
}

export function demoReplyText(messages: ChatMessage[], fx: DemoFixtures = fixtures): string {
  const last = lastUserMessage(messages)
  if (!last) return fx.helpText

  const trimmed = last.content.trim()
  if (trimmed === '/help') return fx.helpText
  if (trimmed.endsWith('?')) return pickFaqAnswer(fx.faqAnswers)
  return fx.echoTemplate.replace('{text}', trimmed)
}

export async function demoReply(
  messages: ChatMessage[],
  options?: { minDelayMs?: number; maxDelayMs?: number; fx?: DemoFixtures },
): Promise<string> {
  const min = options?.minDelayMs ?? 600
  const max = options?.maxDelayMs ?? 1200
  const delay = min + Math.floor(Math.random() * (max - min + 1))
  await new Promise<void>(resolve => setTimeout(resolve, delay))
  return demoReplyText(messages, options?.fx ?? fixtures)
}

File: macos-app/XYZT/Services/DemoAgent.swift
import Foundation

struct DemoFixtures: Decodable {
    let helpText: String
    let faqAnswers: [String]
    let echoTemplate: String
}

enum DemoAgent {
    private static var faqRotation = 0
    private static var cachedFixtures: DemoFixtures?

    static func loadFixtures() -> DemoFixtures {
        if let cached = cachedFixtures { return cached }

        guard let url = Bundle.main.url(forResource: "demo-responses", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let decoded = try? JSONDecoder().decode(DemoFixtures.self, from: data)
        else {
            return DemoFixtures(
                helpText: "XYZT (demo). Use /help for more.",
                faqAnswers: ["Demo FAQ answer."],
                echoTemplate: "Demo: you said «{text}». Real agent wiring comes later."
            )
        }
        cachedFixtures = decoded
        return decoded
    }

    static func resetFaqRotation() {
        faqRotation = 0
    }

    static func pickFaqAnswer(from answers: [String]) -> String {
        guard !answers.isEmpty else { return "" }
        let answer = answers[faqRotation % answers.count]
        faqRotation += 1
        return answer
    }

    static func replyText(for messages: [ChatMessage], fixtures: DemoFixtures? = nil) -> String {
        let fx = fixtures ?? loadFixtures()
        guard let last = messages.last(where: { $0.role == .user }) else {
            return fx.helpText
        }

        let trimmed = last.content.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed == "/help" { return fx.helpText }
        if trimmed.hasSuffix("?") { return pickFaqAnswer(from: fx.faqAnswers) }
        return fx.echoTemplate.replacingOccurrences(of: "{text}", with: trimmed)
    }

    static func reply(for messages: [ChatMessage]) async -> String {
        let delayMs = UInt64.random(in: 600 ... 1200)
        try? await Task.sleep(nanoseconds: delayMs * 1_000_000)
        return replyText(for: messages)
    }
}
